File: 백준/Gold/1062. 가르침/가르침.cc
#include <bits/stdc++.h>
using namespace std;
int n, k, ret, a[54];
string s;

int count(int mask)
{
    int cnt = 0;
    for (int i = 0; i < n; i++)
    {
        if ((a[i] & mask) == a[i])
        {
            cnt++;
        }
    }
    return cnt;
}

int go(int idx, int m, int cnt)
{
    if (cnt > k)
        return 0;
    if (idx == 26)
    {
        return count(m);
    }
    int ret = go(idx + 1, m | (1 << idx), cnt + 1);
    if (idx != 'a' - 'a' && idx != 'n' - 'a' && idx != 't' - 'a' && idx != 'i' - 'a' && idx != 'c' - 'a')
    {
        ret = max(ret, go(idx + 1, m, cnt));
    }
    return ret;
}
int main()
{

    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n >> k;
    for (int i = 0; i < n; i++)
    {
        cin >> s;
        for (char c : s)
        {
            a[i] |= (1 << (c - 'a'));
        }
    }
    cout << go(0, 0, 0) << "\n";

    return 0;
}

File: 백준/Gold/1062. 가르침/가르침.java
import java.util.Scanner;

public class Main {
    static int n, k;
    static int[] a = new int[54];
    static String s;

    static int cal(int visited) {
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (a[i] == (a[i] & visited)) {
                cnt++;
            }
        }
        return cnt;
    }

    static int go(int idx, int cnt, int visited) {
        if (cnt > k) return 0;
        if (idx == 26) return cal(visited);

        int ret = go(idx + 1, cnt + 1, visited | (1 << idx));

        if (idx != 0 && idx != 'n' - 'a' && idx != 't' - 'a' && idx != 'i' - 'a' && 'c' - 'a' != idx) {
            ret = Math.max(ret, go(idx + 1, cnt, visited));
        }

        return ret;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        n = scanner.nextInt();
        k = scanner.nextInt();

        for (int i = 0; i < n; i++) {
            s = scanner.next();
            for (char c : s.toCharArray()) {
                a[i] |= (1 << (c - 'a'));
            }
        }

        System.out.println(go(0, 0, 0));
        scanner.close();
    }
}
